# Modelo Relacional TechCore

## Objetivo

El objetivo es crear un Modelo Relacional con pandas.
Este cuaderno transforma un archivo csv desnormalizado (tabla plana) en un archivo Excel listo para cargar a PowerBI.

### Importación de Librerías y Carga de Datos


In [2]:
# Importar llibrerías
import pandas as pd
import numpy as np

In [3]:
# Cargar archivio csv desde Google Drive

file_id = '1dAoCeJtwqOCaWGQxcn0uOSmgbV0hZrHL'
url = f'https://drive.google.com/uc?id={file_id}'

df = pd.read_csv(url, sep=",")

In [4]:
# Analisis breve de las columnas
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   VentaID                  30000 non-null  int64  
 1   FechaVenta               30000 non-null  object 
 2   HoraVenta                30000 non-null  object 
 3   SucursalNombre           30000 non-null  object 
 4   CiudadSucursal2          30000 non-null  object 
 5   VendedorNombre           30000 non-null  object 
 6   ClienteNombre            30000 non-null  object 
 7   GeneroCliente            30000 non-null  object 
 8   EdadCliente              30000 non-null  int64  
 9   EmailCliente             28437 non-null  object 
 10  TelefonoCliente          29705 non-null  object 
 11  DireccionCliente         30000 non-null  object 
 12  MetodoPago               30000 non-null  object 
 13  NombreProducto1          30000 non-null  object 
 14  MarcaProducto1        

In [5]:
# Mostrar las primeras filas del DataFrame
df.head()

,VentaID,FechaVenta,HoraVenta,SucursalNombre,CiudadSucursal2,VendedorNombre,ClienteNombre,GeneroCliente,EdadCliente,EmailCliente,...,SubtotalProducto2,NombreProducto3,MarcaProducto3,CantidadProducto3,PrecioUnitarioProducto3,SubtotalProducto3,DescuentoVenta,TotalVenta,Year,Month
0,1,12/31/2015,5:42:43,Techcore Pereira,Pereira,Amílcar Ortega-Alberto,Bienvenida Nebot-Fiol,F,37,bienvenida19@hotmail.com,...,16000000.0,Dell Latitude 7420,Dell,1.0,5600000.0,5600000.0,0.0,31200000,2015,12
1,2,3/23/2019,19:03:21,Techcore Medellín #1,Medellín,Ana Sofía Llopis Blázquez,Teófila Bueno-Novoa,F,43,teófila35@gmail.com,...,6800000.0,NaN,NaN,NaN,NaN,NaN,0.0,10800000,2019,3
2,3,12/23/2018,2:32:34,Techcore Medellín #2,Medellín,Juan José Porcel-Riera,Gilberto Chamorro Catalá,M,38,gilberto57@hotmail.com,...,7200000.0,HP Pavilion 15,HP,1.0,3500000.0,3500000.0,0.0,21900000,2018,12
3,4,11/14/2015,0:37:36,Techcore Medellín #2,Medellín,Juan José Porcel-Riera,Máximo Coronado Huerta,M,30,máximo65@hotmail.com,...,8800000.0,NaN,NaN,NaN,NaN,NaN,0.0,13600000,2015,11
4,5,11/29/2016,10:34:20,Techcore Cali,Cali,Jacinta Juárez Marín,Yago Oliver,M,51,yago38@yahoo.com,...,3500000.0,Microsoft Surface Laptop 4,Microsoft,1.0,5200000.0,5200000.0,0.0,13900000,2016,11


### Crear Tablas Normalizadas

En esta estapa definimos las entidades y los atributos para crear las tablas que conformarán la estructura del modelo, a su vez se definen las claves foráneas y primarias.





### Tabla Ciudades

In [6]:
Ciudades =  df['CiudadSucursal2'].unique()
Ciudades = pd.DataFrame(Ciudades, columns=['Ciudad'])
Ciudades['CiudadID'] = Ciudades.index + 1
Ciudades = Ciudades[['CiudadID', 'Ciudad']]
print("---Tabla Ciudades---")
Ciudades


---Tabla Ciudades---


,CiudadID,Ciudad
0,1,Pereira
1,2,Medellín
2,3,Cali
3,4,Bogotá


### Tabla de Sucursales

In [7]:
# Crear la tabla Sucursales
Sucursales = df[['SucursalNombre', 'CiudadSucursal2']].drop_duplicates().reset_index(drop=True)
#Clave primaria SucursalID
Sucursales['SucursalID'] = Sucursales.index + 1
# Agreagamos la llave foranea CiudadID
Sucursales = Sucursales.merge(Ciudades, left_on='CiudadSucursal2', right_on='Ciudad', how='left')
Sucursales = Sucursales[['SucursalID', 'SucursalNombre', 'CiudadID']]
print("---Tabla Sucursales---")
Sucursales

---Tabla Sucursales---


,SucursalID,SucursalNombre,CiudadID
0,1,Techcore Pereira,1
1,2,Techcore Medellín #1,2
2,3,Techcore Medellín #2,2
3,4,Techcore Cali,3
4,5,Techcore Bogotá #2,4
5,6,Techcore Bogotá #1,4


### Tabla Clientes

In [8]:
Clientes = df[['ClienteNombre', 'EmailCliente', 'EdadCliente', 'GeneroCliente', 'TelefonoCliente', 'DireccionCliente']].drop_duplicates().reset_index(drop=True)
#Clave primaria ClienteID
Clientes['ClienteID'] = Clientes.index + 1
Clientes = Clientes[['ClienteID', 'ClienteNombre', 'EmailCliente', 'EdadCliente', 'GeneroCliente', 'TelefonoCliente', 'DireccionCliente']]
print("---Tabla Clientes---")
Clientes.head()

---Tabla Clientes---


,ClienteID,ClienteNombre,EmailCliente,EdadCliente,GeneroCliente,TelefonoCliente,DireccionCliente
0,1,Bienvenida Nebot-Fiol,bienvenida19@hotmail.com,37,F,34806548767,cll 52 #32-98
1,2,Teófila Bueno-Novoa,teófila35@gmail.com,43,F,+34 947255990,cll 96 #81-94
2,3,Gilberto Chamorro Catalá,gilberto57@hotmail.com,38,M,+34 978 810 249,cra 28 #79-85
3,4,Máximo Coronado Huerta,máximo65@hotmail.com,30,M,+34 825429634,cll 5 #89-26
4,5,Yago Oliver,yago38@yahoo.com,51,M,34988285275,cra 32 #97-86


In [9]:
Clientes['ClienteNombre'].value_counts()

ClienteNombre
Leoncio Lledó                2
Lucio del Hurtado            2
Nilo Barrios                 2
Ildefonso Águila             2
Alondra Arias                1
                            ..
Manuel Ibán Cal Cifuentes    1
Adelardo Llamas León         1
Soledad Estévez Bayo         1
Alicia Gelabert Ramos        1
Paco Olivera Suárez          1
Name: count, Length: 17449, dtype: int64

### Tabla Vendedores

In [10]:
Vendedores = df['VendedorNombre'].unique()
Vendedores = pd.DataFrame(Vendedores, columns=['VendedorNombre'])
# Clave primaria VendedorID
Vendedores['VendedorID'] = Vendedores.index + 1
Vendedores = Vendedores[['VendedorID', 'VendedorNombre']]
Vendedores.head()


,VendedorID,VendedorNombre
0,1,Amílcar Ortega-Alberto
1,2,Ana Sofía Llopis Blázquez
2,3,Juan José Porcel-Riera
3,4,Jacinta Juárez Marín
4,5,Clímaco Salinas Zabaleta


### Tabla Método Pago

In [11]:
MetodosPago = df['MetodoPago'].unique()
MetodosPago = pd.DataFrame(MetodosPago, columns=['MetodoPago'])
MetodosPago['MetodoPagoID'] = MetodosPago.index + 1
MetodosPago = MetodosPago[['MetodoPagoID', 'MetodoPago']]
MetodosPago

,MetodoPagoID,MetodoPago
0,1,Tarjeta Crédito
1,2,Billetera Digital
2,3,Tarjeta Débito
3,4,Transferencia
4,5,No Especificado
5,6,Efectivo


### Tabla Facturas


In [12]:
Facturas = df[['VentaID', 'Year', 'Month', 'HoraVenta', 'FechaVenta', 'SucursalNombre',
               'ClienteNombre', 'EmailCliente', 'VendedorNombre', 'MetodoPago',
               'DescuentoVenta','TotalVenta']].drop_duplicates().reset_index(drop=True)

Facturas =  Facturas.merge(Sucursales, left_on='SucursalNombre', right_on='SucursalNombre', how='left')
Facturas = Facturas.merge(Vendedores, left_on='VendedorNombre', right_on='VendedorNombre', how='left')
Facturas = Facturas.merge(Clientes, left_on=['ClienteNombre', 'EmailCliente'], right_on=['ClienteNombre', 'EmailCliente'], how='left')
Facturas = Facturas.merge(MetodosPago, left_on='MetodoPago', right_on ='MetodoPago', how='left')


Facturas = Facturas[['VentaID', 'Year', 'Month', 'HoraVenta', 'FechaVenta', 'SucursalID',
                     'VendedorID', 'ClienteID', 'MetodoPagoID', 'DescuentoVenta','TotalVenta']]

Facturas

,VentaID,Year,Month,HoraVenta,FechaVenta,SucursalID,VendedorID,ClienteID,MetodoPagoID,DescuentoVenta,TotalVenta
0,1,2015,12,5:42:43,12/31/2015,1,1,1,1,0.00,31200000
1,2,2019,3,19:03:21,3/23/2019,2,2,2,2,0.00,10800000
2,3,2018,12,2:32:34,12/23/2018,3,3,3,3,0.00,21900000
3,4,2015,11,0:37:36,11/14/2015,3,3,4,2,0.00,13600000
4,5,2016,11,10:34:20,11/29/2016,4,4,5,1,0.00,13900000
...,...,...,...,...,...,...,...,...,...,...,...
29995,29996,2020,12,1:33:17,12/31/2020,3,3,14207,1,0.00,6800000
29996,29997,2017,7,11:29:20,7/14/2017,4,14,1586,1,0.00,26400000
29997,29998,2019,4,21:38:27,4/6/2019,3,28,17453,4,0.00,5600000
29998,29999,2024,12,6:49:00,12/27/2024,2,25,7089,4,0.00,10800000


### Tabla Productos

In [13]:
# Crear la tabla Productos
Productos = pd.DataFrame(columns=['NombreProducto', 'MarcaProducto', 'Precio'])

# Iteramos sobre los productos en el DataFrame original y los agregamos a la tabla Productos
for x in range (1,4):
    append_producto = pd.DataFrame(columns= ['NombreProducto', 'MarcaProducto', 'Precio'])
    append_producto['NombreProducto'] = df[f'NombreProducto{x}']
    append_producto['MarcaProducto'] = df[f'MarcaProducto{x}']
    append_producto['Precio'] = df[f'PrecioUnitarioProducto{x}']

    Productos = pd.concat([Productos, append_producto], ignore_index=True)

Productos = Productos.dropna().drop_duplicates().reset_index(drop=True)
# Clave primaria ProductoID
Productos['ProductoID'] = Productos.index + 1
Productos = Productos[['ProductoID', 'NombreProducto', 'MarcaProducto', 'Precio']]
print("---Tabla Productos---")
Productos.head()


---Tabla Productos---


,ProductoID,NombreProducto,MarcaProducto,Precio
0,1,Apple MacBook Pro 16,Apple,9600000
1,2,Acer Nitro 5,Acer,4000000
2,3,Dell Latitude 7420,Dell,5600000
3,4,Asus TUF Gaming A15,Asus,4800000
4,5,HP Spectre x360,HP,5200000


### Tabla de Detalles

In [14]:

DetalleFacturas = pd.DataFrame(columns=['VentaID', 'ProductoID', 'Cantidad', 'Subtotal'])

for i in range(1, 4):
    detalle = df[['VentaID', f'NombreProducto{i}',f'MarcaProducto{i}', f'CantidadProducto{i}', f'SubtotalProducto{i}']]
    # Clave foranea ProductoID
    detalle = detalle.merge(Productos, left_on=[f'NombreProducto{i}', f'MarcaProducto{i}'],
                        right_on=['NombreProducto', 'MarcaProducto'], how='left')
    detalle = detalle[['VentaID', 'ProductoID', f'CantidadProducto{i}', f'SubtotalProducto{i}']]
    detalle = detalle.rename(columns={f'CantidadProducto{i}': 'Cantidad', f'SubtotalProducto{i}': 'Subtotal'})
    
    DetalleFacturas = pd.concat([DetalleFacturas, detalle], ignore_index=True)


DetalleFacturas = DetalleFacturas.dropna().reset_index(drop=True)


# DetalleFacturas.info()

# Ajustar tipos de datos
DetalleFacturas['Cantidad'] = DetalleFacturas['Cantidad'].astype(int)
DetalleFacturas['Subtotal'] = DetalleFacturas['Subtotal'].astype(float)
DetalleFacturas['VentaID'] = DetalleFacturas['VentaID'].astype(int)
DetalleFacturas['ProductoID'] = DetalleFacturas['ProductoID'].astype(int)

DetalleFacturas.sort_values(by=['VentaID'], inplace=True)
DetalleFacturas.reset_index(drop=True, inplace=True)

# Clave primaria DetalleID
DetalleFacturas['DetalleID'] = DetalleFacturas.index + 1


DetalleFacturas = DetalleFacturas[['DetalleID', 'VentaID', 'ProductoID', 'Cantidad', 'Subtotal']]

print("---Tabla DetalleFacturas---")
DetalleFacturas.head()



---Tabla DetalleFacturas---


,DetalleID,VentaID,ProductoID,Cantidad,Subtotal
0,1,1,1,1,9600000.0
1,2,1,3,1,5600000.0
2,3,1,22,2,16000000.0
3,4,2,2,1,4000000.0
4,5,2,21,1,6800000.0


## Análisis Exploratorio


### Ventas Por Marca

In [15]:
# Ventas por marca

pd.options.display.float_format = '{:,.2f}'.format

Ventas_por_marca = DetalleFacturas.merge(Productos, on= 'ProductoID')

Ventas_por_marca = Ventas_por_marca.groupby('MarcaProducto')['Subtotal'].sum().reset_index()

Ventas_por_marca = Ventas_por_marca.sort_values(by='Subtotal', ascending=False)

print('---Ventas por marca---')
Ventas_por_marca


---Ventas por marca---


,MarcaProducto,Subtotal
5,Lenovo,"117,879,200,000.00"
4,HP,"96,786,700,000.00"
3,Dell,"86,722,000,000.00"
1,Apple,"78,016,200,000.00"
2,Asus,"33,179,600,000.00"
0,Acer,"22,000,400,000.00"
9,Samsung,"12,760,800,000.00"
6,MSI,"8,525,200,000.00"
7,Microsoft,"8,223,400,000.00"
8,Razer,"3,936,400,000.00"


Lenovo es la marca que más vende.

### Productos Más Vendidos

In [16]:
#10 Productos más vendidos

Ventas_por_producto = DetalleFacturas.merge(Productos, on= 'ProductoID')
Ventas_por_producto = Ventas_por_producto.groupby('NombreProducto')['Subtotal'].sum().reset_index()
Ventas_por_producto = Ventas_por_producto.sort_values(by='Subtotal', ascending=False)
print('---10 Productos más vendidos---')
Ventas_por_producto.head(10)


---10 Productos más vendidos---


,NombreProducto,Subtotal
21,Lenovo Legion 5 Pro,"42,199,200,000.00"
19,HP Spectre x360,"40,175,200,000.00"
22,Lenovo ThinkPad X1 Carbon,"39,535,200,000.00"
17,HP Omen 16,"30,066,000,000.00"
12,Dell Alienware m15,"29,264,000,000.00"
14,Dell Latitude 7420,"25,239,200,000.00"
5,Apple MacBook Pro 14,"24,960,000,000.00"
23,Lenovo Yoga 7i,"22,228,800,000.00"
15,Dell XPS 13,"21,724,800,000.00"
6,Apple MacBook Pro 16,"20,803,200,000.00"


### Distribución de ventas por ciudad

In [17]:
### Distribución de ventas por ciudad

df_analisis =  Facturas.merge(Sucursales, on='SucursalID')

df_analisis = df_analisis.merge(Ciudades, on='CiudadID')

ventas_por_ciudad = df_analisis.groupby('Ciudad')['TotalVenta'].sum().reset_index()

ventas_por_ciudad

total_global = ventas_por_ciudad['TotalVenta'].sum()

ventas_por_ciudad['% Participación'] = (ventas_por_ciudad['TotalVenta'] / total_global) * 100

ventas_por_ciudad = ventas_por_ciudad.sort_values(by='TotalVenta', ascending=False).reset_index(drop=True)

pd.options.display.float_format = '{:,.2f}'.format
print("Distribución de Ventas por Ciudad:")
print(ventas_por_ciudad)



Distribución de Ventas por Ciudad:
     Ciudad    TotalVenta  % Participación
0  Medellín  203861345000            44.66
1    Bogotá  139607510000            30.58
2   Pereira   56619400000            12.40
3      Cali   56403415000            12.36


**Conclusión**

El modelo relacional permite realizar un análisis exploratorio consistente, evitando duplicados o errores de ambiguedad.


**Distribución de ventas por ciudad**

- Medellín es la ciudad líder con $203,861 millones (un contundente 44.66% de todas las ventas).

- Bogotá ocupa el segundo lugar con el 30.58%.

- Pereira y Cali están muy empatadas en el tercer puesto con aproximadamente un 12.4% cada una.

## Exportar Modelo a Archivo Excel

In [16]:
dataframes = [
    Ciudades,
    Sucursales,
    Clientes,
    Vendedores,
    MetodosPago,
    Facturas,
    Productos,
    DetalleFacturas
]


nombrar_hojas = [
    'Ciudades',
    'Sucursales',
    'Clientes',
    'Vendedores',
    'MetodosPago',
    'Facturas',
    'Productos',
    'DetalleFacturas'
]


for df, nombre_hoja in zip(dataframes, nombrar_hojas):
    df.fillna(' ', inplace=True)

ruta_archivo = 'ModeloVentas.xlsx'

with pd.ExcelWriter(ruta_archivo, engine='xlsxwriter') as writer:
    for df, nombre_hoja in zip(dataframes, nombrar_hojas):
        df.to_excel(writer, sheet_name=nombre_hoja, index=False)
